[Shee Hui] iP

sheehui opened the pull request:
# Duke

> Someday you will be old enough to read fairytales again ~C. S. Lewis ([more](https://www.reddit.com/r/books/comments/1jiop4/what_does_the_cs_lewis_quote_some_day_you_will_be/))

Duke frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- **super fast**

All you need to do it:
1. download it
2. run it in your local environment
3. add your tasks
4. start working towards your dreams!

It is all ***FREE***!!!! :confetti_ball: :partying_face: :tada:

Features:
- [X] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

sheehui committed:
Amend gitignore file

Add *.class

sheehui committed:
Complete Level-1

Add skeletal feature to greet user, echo command and exit on 'bye'

sheehui committed:
Complete Level-2

Add function to store and add task + list tasks

sheehui committed:
Complete Level-3

Allow users to mark task as done + add Task class

sheehui committed:
Complete Level-4

Add 3 types of tasks (todo, deadline, event), which inherits from Task class.

sheehui committed:
Complete A-TextUiTesting

Amend runtest.sh to compare output + update actual/expected tests

sheehui committed:
Complete Level-5

Handle bad user inputs with exceptions and thrown errors.

sheehui committed:
Complete Level-6

Change taskList from an array to an arraylist. Added delete task function. Fix string concatenation.

sheehui committed:
Run prettier

sheehui committed:
Add checkstyle file

sheehui committed:
Change static methods to non-static

Created a Duke instance so methods can be non-static

sheehui committed:
Add enum class for Task types

Amend Duke class to add Task by recognising enum types.

sheehui committed:
Save tasks to local db

Add a Storage class that allows Duke to read from a db .txt file
when initialised and write to a db file when stopped.

sheehui committed:
Store dates as java.time.LocalDate

sheehui committed:
Merge branch 'branch-Level-7'

sheehui committed:
Resolve merge conflict

sheehui committed:
Add TaskList for OOP

sheehui committed:
Add Parser class for OOP

Process user command and returns an array containing
input information.

sheehui committed:
Add Ui class for OOP

Deals with interactions with the user. i.e. add task, delete, finish
etc.

sheehui committed:
Complete A-Packages

* duke for Duke
* duke.exception for all exception classes
* duke.task for all task related classes
* duke.util for Ui, Parser and Storage

sheehui committed:
Fix bug in separateDetails

Throw error if there is nothing before and after /at, /by.

sheehui committed:
Amend task classes to handle error with date inputs

sheehui committed:
Add JUnit tests

Test deadline, event, todo for:
-output string
-error thrown for wrong datetime
-correct taskToString conversion

sheehui committed:
Package the app as a JAR file

sheehui committed:
Complete Level-9

Find a task by searching for a keyword.

sheehui committed:
Complete JavaDocs.

Write java documentation for public classes/methods and some private
methods.

sheehui committed:
Configure project to match the necessary coding standards

sheehui committed:
Merge branch 'branch-Level-9'

sheehui committed:
Merge branch 'branch-A-JavaDoc'

Resolve conflicts from adding 'find' function.

sheehui committed:
Merge branch 'branch-A-CodingStandard'

Resolve conflicts and add Javadocs for find related functions.

sheehui committed:
Merge branch 'add-gradle-support'

sheehui committed:
Run checkstyle in util package.

Fix style errors.

sheehui committed:
Run checkstyle in exception package

Fix style errors.

sheehui committed:
Run checkstyle in tasks package

Fix style errors.

sheehui committed:
Merge branch 'branch-A-CheckStyle'

sheehui committed:
Implement GUI for Duke

sheehui committed:
Implement command package

Assign each command type a command class that has an execute function.

sheehui committed:
Amend parser, ui and others to accomodate Command package

Add more oop to Duke project.

sheehui committed:
Add CSS to GUI + javadocs

sheehui committed:
Merge branch 'branch-Level-10'

sheehui committed:
Write to storage at each successful task update

Remove storage as a parameter in command class and call static storage method on each write.

sheehui committed:
Fix bug in Label height

All output renders on GUI instead of being truncated.

sheehui committed:
Make UserInputError the only exception class

Removed other exception classes.

tau-bar commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public abstract class Command {
+    /**
+     * Executes the user's command.
+     * @param tasks Current tasklist.
+     * @param ui User interface of Duke.
```
I think you should include an empty line between the description and parameter section based on [this](https://se-education.org/guides/conventions/java/intermediate.html#comments).

```suggestion
     * Executes the user's command.
     * 
     * @param tasks Current tasklist.
     * @param ui User interface of Duke.
```

tau-bar commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+    protected LocalDate when;
+
+    /**
+     * Constructor to create a Dedline task.
```
```suggestion
     * Constructor to create a Deadline task.
```

tau-bar commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,48 @@
+     * @param done Boolean value that indicates completeness of task.
+     * @throws UserInputError
+     */
+    public Event(String description, String when, boolean done) throws UserInputError {
+        super(description, Task.Type.EVENT, done);
+        try {
+            this.when = LocalDate.parse(when.trim());
+        } catch (DateTimeException e) {
+            throw new NoDescriptionException("Wrong datetime format");
+        }
+    }
```
I think that the parameter naming can be named to sound more like a boolean.
```suggestion
    public Event(String description, String when, boolean isEventDone) throws UserInputError {
        super(description, Task.Type.EVENT, isEventDone);
        try {
            this.when = LocalDate.parse(when.trim());
        } catch (DateTimeException e) {
            throw new NoDescriptionException("Wrong datetime format");
        }
    }
```

tau-bar commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,197 @@
+     * @param input String containing task details.
+     * @param type The enum type of the task.
+     * @return A Task object.
+     * @throws UserInputError
```
I think you can put the condition when the error is thrown
i.e. ```* @throws UserInputError If ...```

tau-bar commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * Return a formated icon to symbolise if task is completed.
```
```suggestion
     * Return a formatted icon to symbolize if task is completed.
```

tau-bar commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+        taskList = list;
+    }
+
+    /**
+     * Find the number of Tasks.
+     *
+     * @return
+     */
+    public int length() {
+        return taskList.size();
+    }
```
I think you should put what the method returns after the ```@return``` tag.
```suggestion
    /**
     * Find the number of Tasks.
     *
     * @return The number of tasks in the task list.
     */
    public int length() {
        return taskList.size();
    }
```

tau-bar commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,158 @@
+package duke.util;
+
+import duke.Duke;
+import duke.commands.*;
```
I think you should explicitly state the imports as stated [here.](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

tau-bar reviewed (commented):
LGTM, however I noticed some minor coding standard violations, which I have pointed out. I'm impressed at the depth of documentation in your code, and the abstraction of each Command into their own class. Good job! 👍🏽

sheehui committed:
Remove exception classes and update styling

sheehui committed:
Update fat jar file with gradle

Use shadowJar to wrap javafx application.

sheehui committed:
Merge branch 'Add-ShadowJar'

sheehui committed:
Add checkstyle config

Use gradle to perform checkstyle instead of IDEA-plugin.

sheehui committed:
Add assert to project

Check that description is not empty and index references are valid for "done" and "delete" commands.

sheehui committed:
Improve code quality

Make changes to method names for better coherence.

sheehui committed:
Merge pull request #5 from sheehui/branch-A-Assertions

Add assert

sheehui committed:
Merge branch 'master' into branch-A-CodeQuality

sheehui committed:
Fix bugs from merging updated master

Change taskList method length to getNumOfTasks.

sheehui committed:
Merge pull request #4 from sheehui/branch-A-CodeQuality

Improve code quality

sheehui committed:
Add extension for Duke to accept natural dates

Event and Deadline tasks now support natural date formats such as "mon", "tues"...."sun", "today", "tmr"

Users have the option to use natural dates in place of yyyy-mm-dd format when their input task is within 7 days from the current date.

getDateFromString method is added to the parent class Task for Event and Deadline to inherit, to avoid code duplication.

sheehui committed:
Add javadocs to new method

A new method getDateFromString was added to allow users to input natural dates.

sheehui committed:
Merge pull request #6 from sheehui/branch-BCD-Extension

Add support for natural dates as input for event and deadline tasks.

sheehui committed:
Add CI and github workflow to project

sheehui committed:
Add .yml file for CI

sheehui committed:
Appease checkstyle errors

sheehui committed:
Fix CustomImportError my moving duke imports

sheehui committed:
Fix JUnit tests + import order errors

sheehui committed:
Merge pull request #7 from sheehui/github-actions

Add CI and github workflow to project

sheehui committed:
Add time for deadline tasks

Users now can input date and time into deadline tasks with a whitespace between the two.
A check function ensures that time is in the correct format + is a valid time.

sheehui committed:
Appease checkstyle errors + Fix deadline tests

Fix errors arising from adding time function to deadline.

sheehui committed:
Make duke user and bot pictures round.

Clipped ImageView to the shape of a circle.

sheehui committed:
Clean up code

Remove redundant code and make code look neater.

sheehui committed:
Update GUI

sheehui committed:
Update event and deadline task to accept time as an input.

Deadline now has time as an optional input for users.
Event now requires both date and time to be created, time includes starting and ending time with a dash to separate them.

sheehui committed:
Update User Guide

Add feature list details for:
1. task manager
2. find function
3. natural dates support

Add Usage guide for adding, deleting, completing, viewing tasks. Explain how to find tasks containing a specific keyword and how to exit Duke.

sheehui committed:
Fix JUnit tests

Tests failed due to amendment of datetime formats in deadline and event tasks.

sheehui committed:
Merge pull request #9 from sheehui/week-6

Create user guide and improve GUI

sheehui committed:
Update README

Remove emojis to ensure GitHub pages is rendered correctly

sheehui committed:
Upload representative screenshot of Duke

sheehui committed:
Update shadow jar to enable Duke in linux and windows

sheehui committed:
Add format of each command to the user guide and fix mistakes

Show input format to users so they can use Duke more effectively.

sheehui committed:
Amend output format of Duke for taskList, error messages + handle whitespaces for commands/datetime

Enable smoother user experience.

sheehui committed:
Add message to assertions

To promote easier debugging.

sheehui committed:
Final amend to Duke output

sheehui committed:
Update javaFX version to 12

project kept getting an assertion failure that came from a bug in the dependency.

sheehui committed:
Update README expected output to reflect changes in Duke

sheehui committed:
Update Ui.png

Made changes to Duke output

sheehui committed:
Update JUnit tests

Amend test expected output due to updates in Duke output.

sheehui committed:
Improve javadoc comments

Amend javadocs according to bot feedback.

sheehui committed:
Override javadoc comments for Command subclasses

sheehui committed:
Add function to respond to empty inputs

False command checks for empty strings and generates an empty input error message.

sheehui committed:
Update javaFX version in fxml

sheehui committed:
Add Javadocs to public methods in ui package for the GUI

sheehui committed:
Merge pull request #10 from sheehui/final-branch

Final branch
